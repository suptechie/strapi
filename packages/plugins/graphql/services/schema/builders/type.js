'use strict';

const { isArray, isString, isUndefined } = require('lodash/fp');
const { objectType } = require('nexus');

const { contentTypes } = require('@strapi/utils');

const { mappers, utils: typeUtils, constants } = require('../../types');
const { buildAssociationResolver, buildComponentResolver } = require('../resolvers');
const { getContentTypeArgs } = require('./utils');

/**
 * @typedef TypeBuildersOptions
 *
 * @property {ObjectDefinitionBlock} builder
 * @property {string} attributeName
 * @property {object} attribute
 * @property {object} contentType
 * @property {object} context
 * @property {object} context.strapi
 * @property {object} context.registry
 */

module.exports = context => ({
  /**
   * Create a type definition for a given content type
   * @param contentType - The content type used to created the definition
   * @return {NexusObjectTypeDef}
   */
  buildTypeDefinition(contentType) {
    const { attributes, modelType, options = {} } = contentType;

    const attributesKey = Object.keys(attributes);
    const hasTimestamps = isArray(options.timestamps);

    const name = (modelType === 'component'
      ? typeUtils.getComponentName
      : typeUtils.getTypeName
    ).call(null, contentType);

    return objectType({
      name,

      definition(t) {
        // 1. Timestamps
        // If the content type has timestamps enabled
        // then we should add the corresponding attributes in the definition
        if (hasTimestamps) {
          const [createdAtKey, updatedAtKey] = contentType.options.timestamps;

          t.nonNull.dateTime(createdAtKey);
          t.nonNull.dateTime(updatedAtKey);
        }

        /** 2. Attributes
         *
         * Attributes can be of 7 different kind:
         * - Scalar
         * - Component
         * - Dynamic Zone
         * - Enum
         * - Media
         * - Polymorphic Relations
         * - Regular Relations
         *
         * Here, we iterate over each non-private attribute
         * and add it to the type definition based on its type
         */
        attributesKey
          // Ignore private attributes
          .filter(isNotPrivate(contentType))
          // Add each attribute to the type definition
          .forEach(attributeName => {
            const attribute = attributes[attributeName];

            // We create a copy of the builder (t) to apply custom
            // rules only on the current attribute (eg: nonNull, list, ...)
            let builder = t;

            if (attribute.required) {
              builder = builder.nonNull;
            }

            /**
             * @type {TypeBuildersOptions}
             */
            const options = { builder, attributeName, attribute, contentType, context };

            // Scalars
            if (typeUtils.isStrapiScalar(attribute)) {
              addScalarAttribute(options);
            }

            // Components
            else if (typeUtils.isComponent(attribute)) {
              addComponentAttribute(options);
            }

            // Dynamic Zones
            else if (typeUtils.isDynamicZone(attribute) && attribute.components.length > 0) {
              addDynamicZoneAttribute(options);
            }

            // Enums
            else if (typeUtils.isEnumeration(attribute)) {
              addEnumAttribute(options);
            }

            // Media
            else if (typeUtils.isMedia(attribute)) {
              addMediaAttribute(options);
            }

            // Polymorphic Relations
            else if (typeUtils.isMorphRelation(attribute)) {
              addPolymorphicRelationalAttribute(options);
            }

            // Regular Relations
            else if (typeUtils.isRelation(attribute) || typeUtils.isMedia(attribute)) {
              addRegularRelationalAttribute(options);
            }
          });
      },
    });
  },
});

/**
 * Add a scalar attribute to the type definition
 *
 * The attribute is added based on a simple association between a Strapi
 * type and a GraphQL type (the map is defined in `strapiTypeToGraphQLScalar`)
 *
 * @param {TypeBuildersOptions} options
 */
const addScalarAttribute = ({ builder, attributeName, attribute }) => {
  const gqlType = mappers.strapiScalarToGraphQLScalar(attribute.type);

  builder.field(attributeName, { type: gqlType });
};

/**
 * Add a component attribute to the type definition
 *
 * The attribute is added by fetching the component's type
 * name and using it as the attribute's type
 *
 * @param {TypeBuildersOptions} options
 */
const addComponentAttribute = ({ builder, attributeName, contentType, attribute, context }) => {
  const { strapi } = context;

  const type = typeUtils.getComponentNameFromAttribute(attribute);

  if (attribute.repeatable) {
    builder = builder.list;
  }

  const targetComponent = strapi.getModel(attribute.component);

  const resolve = buildComponentResolver({
    contentTypeUID: contentType.uid,
    attributeName,
    strapi,
  });

  const args = getContentTypeArgs(targetComponent);

  builder.field(attributeName, { type, resolve, args });
};

/**
 * Add a dynamic zone attribute to the type definition
 *
 * The attribute is added by fetching the dynamic zone's
 * type name and using it as the attribute's type
 *
 * @param {TypeBuildersOptions} options
 */
const addDynamicZoneAttribute = ({ builder, attributeName, contentType }) => {
  const type = typeUtils.getDynamicZoneName(contentType, attributeName);

  builder.field(attributeName, { type });
};

/**
 * Add an enum attribute to the type definition
 *
 * The attribute is added by fetching the enum's type
 * name and using it as the attribute's type
 *
 * @param {TypeBuildersOptions} options
 */
const addEnumAttribute = ({ builder, attributeName, contentType }) => {
  const type = typeUtils.getEnumName(contentType, attributeName);

  builder.field(attributeName, { type });
};

/**
 * Add a media attribute to the type definition
 * @param {TypeBuildersOptions} options
 */
const addMediaAttribute = options => {
  let { builder } = options;
  const {
    attributeName,
    attribute,
    contentType,
    context: { strapi },
  } = options;

  const fileContentType = strapi.getModel('plugins::upload.file');

  const resolve = buildAssociationResolver({
    contentTypeUID: contentType.uid,
    attributeName,
    strapi,
  });

  const args = attribute.multiple ? getContentTypeArgs(fileContentType) : undefined;
  const type = attribute.multiple
    ? typeUtils.getEntityResponseCollectionName(fileContentType)
    : typeUtils.getEntityResponseName(fileContentType);

  builder.field(attributeName, { type, resolve, args });
};

/**
 * Add a polymorphic relational attribute to the type definition
 * @param {TypeBuildersOptions} options
 */
const addPolymorphicRelationalAttribute = options => {
  let { builder } = options;
  const {
    attributeName,
    attribute,
    contentType,
    context: { strapi },
  } = options;

  const { target } = attribute;
  const isToManyRelation = attribute.relation.endsWith('Many');

  if (isToManyRelation) {
    builder = builder.list;
  }
  // todo[v4]: How to handle polymorphic relation w/ entity response collection types?
  //            -> Currently return raw polymorphic entities

  const resolve = buildAssociationResolver({
    contentTypeUID: contentType.uid,
    attributeName,
    strapi,
  });

  // If there is no specific target specified, then use the GenericMorph type
  if (isUndefined(target)) {
    builder.field(attributeName, {
      type: constants.GENERIC_MORPH_TYPENAME,
      resolve,
    });
  }

  // If the target is an array of string, resolve the associated morph type and use it
  else if (isArray(target) && target.every(isString)) {
    const type = typeUtils.getMorphRelationTypeName(contentType, attributeName);

    builder.field(attributeName, { type, resolve });
  }
};

/**
 * Add a regular relational attribute to the type definition
 * @param {TypeBuildersOptions} options
 */
const addRegularRelationalAttribute = options => {
  let { builder } = options;
  const {
    attributeName,
    attribute,
    contentType,
    context: { strapi },
  } = options;

  const isToManyRelation = attribute.relation.endsWith('Many');

  const resolve = buildAssociationResolver({
    contentTypeUID: contentType.uid,
    attributeName,
    strapi,
  });

  const targetContentType = strapi.getModel(attribute.target);

  const type =
    isToManyRelation && targetContentType.kind !== 'singleType'
      ? typeUtils.getEntityResponseCollectionName(targetContentType)
      : typeUtils.getEntityResponseName(targetContentType);

  const args = isToManyRelation ? getContentTypeArgs(targetContentType) : undefined;

  builder.field(attributeName, { type, resolve, args });
};

/**
 * Bind a content type on an attribute privacy checker
 *
 * @param {object} contentType
 * @return {function(string): boolean}
 */
const isNotPrivate = contentType => attributeName => {
  return !contentTypes.isPrivateAttribute(contentType, attributeName);
};
