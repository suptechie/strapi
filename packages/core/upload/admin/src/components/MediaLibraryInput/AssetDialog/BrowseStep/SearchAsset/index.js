import React, { useState, useLayoutEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import { Searchbar, SearchForm } from '@strapi/design-system/Searchbar';
import { IconButton } from '@strapi/design-system/IconButton';
import SearchIcon from '@strapi/icons/Search';
import getTrad from '../../../../../utils/getTrad';

const SearchAsset = ({ onChangeSearch, onClearSearch, onSubmitSearch, searchValue }) => {
  const { formatMessage } = useIntl();
  const [isOpen, setIsOpen] = useState(!!searchValue);
  const wrapperRef = useRef(null);

  useLayoutEffect(() => {
    if (isOpen) {
      setTimeout(() => {
        wrapperRef.current.querySelector('input').focus();
      }, 0);
    }
  }, [isOpen]);

  const handleToggle = () => {
    setIsOpen(prev => !prev);
  };

  const handleClear = () => {
    handleToggle();
    onClearSearch();
  };

  if (isOpen) {
    return (
      <div ref={wrapperRef}>
        <SearchForm onSubmit={onSubmitSearch}>
          <Searchbar
            name="search"
            onClear={handleClear}
            onChange={e => onChangeSearch(e.target.value)}
            clearLabel={formatMessage({
              id: getTrad('search.clear.label'),
              defaultMessage: 'Clear the search',
            })}
            size="S"
            value={searchValue}
            placeholder={formatMessage({
              id: getTrad('search.placeholder'),
              defaultMessage: 'e.g: the first dog on the moon',
            })}
          >
            {formatMessage({ id: getTrad('search.label'), defaultMessage: 'Search for an asset' })}
          </Searchbar>
        </SearchForm>
      </div>
    );
  }

  return <IconButton icon={<SearchIcon />} label="Search" onClick={handleToggle} />;
};

SearchAsset.propTypes = {
  onChangeSearch: PropTypes.func.isRequired,
  onClearSearch: PropTypes.func.isRequired,
  onSubmitSearch: PropTypes.func.isRequired,
  searchValue: PropTypes.string.isRequired,
};

export default SearchAsset;
