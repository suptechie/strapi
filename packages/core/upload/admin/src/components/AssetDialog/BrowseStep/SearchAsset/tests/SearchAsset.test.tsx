// TODO: find a better naming convention for the file that was an index file before
import { DesignSystemProvider } from '@strapi/design-system';
import { fireEvent, render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { SearchAsset } from '../SearchAsset';

import type { Query } from '../../../../../../../shared/contracts/files';

const handleChange = jest.fn();

const makeApp = (queryValue: Query['_q'] | null) => (
  <DesignSystemProvider>
    <IntlProvider locale="en">
      <SearchAsset onChangeSearch={handleChange} queryValue={queryValue} />
    </IntlProvider>
  </DesignSystemProvider>
);

describe('SearchAsset', () => {
  it('renders and matches the snapshot', () => {
    const { container } = render(makeApp(null));

    expect(container).toMatchSnapshot();
  });

  it('should set input value to queryValue if it exists', () => {
    const queryValue = 'michka';
    const { getByRole } = render(makeApp(queryValue));

    const input = getByRole('textbox', {
      name: 'search',
    }) as HTMLInputElement;

    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual(queryValue);
  });

  it('should call handleChange when submitting search input', () => {
    const { getByRole } = render(makeApp(null));

    const button = getByRole('button');

    if (button) {
      fireEvent.click(button);
    }
    const input = getByRole('textbox', {
      name: 'search',
    });

    if (input) {
      fireEvent.change(input, { target: { value: 'michka' } });
      fireEvent.submit(input);
    }

    expect(handleChange.mock.calls.length).toBe(1);
  });
});
