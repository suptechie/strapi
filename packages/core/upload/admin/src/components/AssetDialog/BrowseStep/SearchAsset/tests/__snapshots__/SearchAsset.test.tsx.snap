// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAsset renders and matches the snapshot 1`] = `
.c0 {
  border-radius: 4px;
  display: inline-flex;
  cursor: pointer;
}

.c1 {
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: flex;
}

.c3 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  text-decoration: none;
  padding-block: 0.7rem;
  padding-inline: 0.7rem;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
  color: #8e8ea9;
}

.c2:hover {
  background-color: #f6f6f9;
  color: #666687;
}

.c2:active {
  background-color: #eaeaef;
}

.c2[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

@media (prefers-reduced-motion: no-preference) {
  .c2 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

<div>
  <button
    aria-disabled="false"
    class="c0 c1 c2"
    data-state="closed"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      focusable="false"
      height="16"
      viewBox="0 0 32 32"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M29.061 26.939 23.125 21A11.515 11.515 0 1 0 21 23.125l5.941 5.942a1.503 1.503 0 0 0 2.125-2.125zM5.5 14a8.5 8.5 0 1 1 8.5 8.5A8.51 8.51 0 0 1 5.5 14"
      />
    </svg>
    <span
      class="c3"
    >
      Search
    </span>
  </button>
  <span
    class="c3"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
